File: src/1-500/002/AddTwoNumbers.py
from functools import reduce

class Node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next

    def __str__(self):
        if self.next:
            return '%s, %s' % (self.value, self.next)
        return '%s' % self.value


class Solution:
    def addThreeNumbers(self, num0: Node, num1: Node, num2: Node, newList: Node) -> Node:
        if num1 is None and \
                num2 is None and \
                num0 is None:
            return 

        value = ( num0.value if num0 is not None else 0 ) + \
                ( num1.value if num1 is not None else 0 ) + \
                ( num2.value if num2 is not None else 0 )
        newNode = Node(value % 10)
        newList.next = newNode

        incNode = Node(value//10) if value // 10 != 0 else None

        self.addThreeNumbers(
            incNode,
            num1.next if num1 is not None else None,
            num2.next if num2 is not None else None,
            newNode)

    def addTwoNumbers(self, num1: Node, num2: Node) -> Node:
        res = Node(-1)
        self.addThreeNumbers(None, num1, num2, res)
        return res.next


if __name__ == '__main__':
    nums1 = [2, 7, 6, 5]
    nums2 = [2, 7, 7, 5]

    def bindNode(pre, num):
        node = Node(num)
        pre.next = node
        return node

    ns1 = Node(-1)
    reduce(bindNode, nums1, ns1)
    ns1 = ns1.next

    ns2 = Node(-1)
    reduce(bindNode, nums2, ns2)
    ns2 = ns2.next

    res = Solution().addTwoNumbers(ns1, ns2)
    print('ns1: ', ns1)
    print('ns2: ', ns2)
    print('res: ', res)

File: src/1-500/002/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode sum = new ListNode(-1);
        ListNode cur = sum;
        int tens = 0;
        while (l1!= null && l2 != null) {
            int n1 = l1.val;
            int n2 = l2.val;
            
            int n = n1 + n2 + tens;
            if (n >= 10) {
                tens = 1;
                n -= 10;
            }
            else {
                tens = 0;
            }
            
            cur.next = new ListNode(n);
            cur = cur.next;
            l1 = l1.next;
            l2 = l2.next;
        }
        
        while (l1!= null) {
            int n1 = l1.val;
            int n = n1 + tens;
            if (n >= 10) {
                tens = 1;
                n -= 10;
            }
            else {
                tens = 0;
            }
            
            cur.next = new ListNode(n);
            cur = cur.next;
            l1 = l1.next;
        }
        
        while (l2!= null) {
            int n2 = l2.val;
            int n = n2 + tens;
            if (n >= 10) {
                tens = 1;
                n -= 10;
            }
            else {
                tens = 0;
            }
            
            cur.next = new ListNode(n);
            cur = cur.next;
            l2 = l2.next;
        }
        
        if (tens > 0) {
            cur.next = new ListNode(1);
        }
        
        return sum.next;
    }
}